Hopefully restore aspnet-request-posted-body using middleware approac…

bakgerman opened the pull request:
…h and HttpContext.Items collection

bakgerman commented:
Attempt to restore aspnet-request-posted-body layout renderer in NLog.Web 5.0.

snakefoot commented:
Thank you for this great contribution. Really amazing that you include both NLog.Web (HttpModule) and NLog.Web.AspNetCore (Middleware). I like the idea of using delegates, that allows the users to plugin functionality without having to inherit and override methods.

Would it make sense if we created one AspNetCore-middleware for NLog, where one of the features is correctly capturing `${aspnet-request-posted-body}` ? Was thinking to merge it together with NLogRequestLoggingMiddleware, so request-body capture is one extra feature (And change NLogRequestLoggingMiddleware so its current exception-/error-logging also become plugin-style).

Was also thinking to make the Enable-Buffering-Condition-Logic a little "smarter", since the `EnableBuffering()`-buffering is super expensive. So maybe check the ContentType / ContentLength before actually doing the buffering. And if the buffering is below 64 KByte then just use a byte-array to avoid using the heavy tmp-file-buffering implemented by `EnableBuffering()`

bakgerman commented:
Thank you, I worked very hard on the PR.  :)

If you want a single middleware class that is clearly the proper goal, I did not want to change an existing class since this is my first commit to NLog.Web, but yes i thought when I was doing the PR that having > 1 middleware is not the best idea.
Can you please explain 'become plugin style' in some more detail before I feel I can make the change?

I can do the EnableBuffering() only if < 64K change quickly, later today or tomorrow. 
I will have to test it thru an actual http flow as I recall sometimes bad things happened if this was off, but that ight have been only for > 64K case.

I have used NLog heavily and would like to help enhance one of my favorite github projects.

snakefoot commented:
> If you want a single middleware class that is clearly the proper goal, I did not want to change an existing class since this is my first commit to NLog.Web

I was also guessing that was your approach (to keep things simple and isolated before interfering with existing classes).

> Can you please explain 'become plugin style' in some more detail before I feel I can make the change?

You have created `NLogRequestPostedBodyMiddlewareConfiguration` where one can plugin custom delegates, that will be used instead of the default-delegate.

I was thinking a similar thing could be done for `NLogRequestLoggingMiddleware`, so it calls a plugin-method defined on `NLogRequestLoggingOptions` that decides whether to do exception-/error-logging.

bakgerman commented:
```
"You have created NLogRequestPostedBodyMiddlewareConfiguration where one can plugin custom delegates, that will be used instead of the default-delegate.

I was thinking a similar thing could be done for NLogRequestLoggingMiddleware, so it calls a plugin-method defined on NLogRequestLoggingOptions that decides how it should perform exception-/error-logging (when assigned null then it is skipped)."
```

I see what you mean now.

For the EnableBuffering(), it seems like Microsoft has already done the in memory part for us:
"Normally buffers request bodies in memory; writes requests larger than 30K bytes to disk."

As I am using my middleware for rest and soap, i do not have a test case > 30K.
We did test the performance of the middleware and async NLog.Database logging, with the commercial DynaTrace tool, and it was found to not harm any performance.

snakefoot commented:
> For the EnableBuffering(), it seems like Microsoft has already done the in memory part for us:
> > "Normally buffers request bodies in memory; writes requests larger than 30K bytes to disk."

Then I think 30K should be the default-max-limit for NLog-midddleware, and if ContentLength is larger, then it will not attempt `EnableBuffering()`.

bakgerman commented:
Well, I found that if I leave EnableBuffing() off, stream.CanSeek returns false and we cannot capture the request body since we have to reset Position property to zero, read to end, then rest the Position back to original.  The unit test does not show this but actual end to end test does show this.

bakgerman commented:
ok I will make limit = 30k instead of 8 k

bakgerman commented:
Default limit changed from 8KB to 30KB per Microsoft comments on EnableBuffering() method.

bakgerman commented:
Will need help testing the IHttpModule version end to end, I do not have any legacy 4.6.2 Framework web application anymore at my team where I work.

bakgerman commented:
Will examine squeezing 2 middleware into 1 class later this week.

bakgerman commented:
I think I am ready for the next review.  The last commit unit tests failed due to an unrelated nuget.org error and I am researching how to rerun the checks for the last commit.

snakefoot commented on src/NLog.Web.AspNetCore/NLogRequestPostedBodyMiddleware.cs:
```diff
@@ -0,0 +1,138 @@
+            }
+
+            // This is required, otherwise reading the request will destructively read the request
+            context.Request.EnableBuffering();
```
I think `EnableBuffering` should be moved inside `if (_configuration.ShouldCapture(context))`

bakgerman replied:
Completed as requested

snakefoot commented on src/NLog.Web/NLogRequestPostedBodyHttpModule.cs:
```diff
@@ -0,0 +1,169 @@
+
+            if (shouldCapture)
+            {
+                items[AspNetRequestPostedBodyLayoutRenderer.NLogPostedRequestBodyKey] = GetString(bodyStream);
```
When `GetString(bodyStream)` returns `null`  or `string.Empty`, then it should not insert into items

bakgerman replied:
Completed as requested

snakefoot commented on src/NLog.Web.AspNetCore/NLogRequestPostedBodyMiddleware.cs:
```diff
@@ -0,0 +1,138 @@
+        public async Task InvokeAsync(HttpContext context, RequestDelegate next)
+        {
+            // Perform null checking
+            if (context.Request == null)
```
Could all these `if`-statements be moved into a private-methods, so it becomes like this:

```c#
if (!ShouldCaptureRequestBody(context))
{
    await next(context).ConfigureAwait(false);
    return;
}
```

bakgerman replied:
Completed as requested

snakefoot commented on src/NLog.Web.AspNetCore/NLogRequestPostedBodyMiddleware.cs:
```diff
@@ -0,0 +1,138 @@
+            {
+                // Save the POST request body in HttpContext.Items with a key of '__nlog-aspnet-request-posted-body'
+                context.Items[AspNetRequestPostedBodyLayoutRenderer.NLogPostedRequestBodyKey] =
+                    await GetString(context?.Request.Body).ConfigureAwait(false);
```
When `GetString(bodyStream)` returns null or string.Empty, then it should not insert into items

bakgerman replied:
completed as requested

snakefoot commented on src/NLog.Web/NLogRequestPostedBodyHttpModule.cs:
```diff
@@ -0,0 +1,169 @@
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="args"></param>
+        protected void InterceptRequest(object sender, EventArgs args)
```
Could this method become private and renamed to `BeginRequest`. I prefer to keep the visibility of methods to absolute minimum, until someone needs access.

bakgerman replied:
Completed as requested

snakefoot commented on src/NLog.Web/NLogRequestPostedBodyHttpModule.cs:
```diff
@@ -0,0 +1,169 @@
+        {
+            HttpApplication app = sender as HttpApplication;
+
+            CaptureRequestPostedBody(
```
Could this become

```c#
if (ShouldCaptureRequestPostedBody(app))
{
      CaptureRequestPostedBody(app);
}
```

bakgerman replied:
I am working on this today.

bakgerman replied:
I finished this, but encountered problems when unit testing.  I then make the private methods internal, and try to use InternalsVisibleTo attribute, but was unable to find the proper value for the Public Key.  The HttpContext and HttpRequest are not mockable in .NET 3.x and 4.x as they are sealed, and NSubstitute throws an exception attempting to mock them, that is why i tried to make the private methods internal so I can pass them arguments like MemoryStream directly from the unit tests.  The unit tests for this class are currently commented until this is resolved.

snakefoot replied:
If it hurts unit-test coverage, then just keep them public, then I can try and figure out why InternalsVisibleTo is not working.

snakefoot commented on src/NLog.Web/NLogRequestPostedBodyMiddlewareConfiguration.cs:
```diff
@@ -0,0 +1,50 @@
+        /// </summary>
+        public static bool DefaultCapture(HttpApplication app)
+        {
+            return app?.Context?.Request?.ContentLength != null && app?.Context?.Request?.ContentLength <= 30 * 1024;
```
Could `30 * 1024` become a setting ?

bakgerman replied:
Made a public property named MaximumRequestSize defaulting to 30KB.

snakefoot commented on src/NLog.Web/NLogRequestPostedBodyMiddlewareConfiguration.cs:
```diff
@@ -0,0 +1,50 @@
+        /// <summary>
+        /// Defaults to 1024
+        /// </summary>
+        public int BufferSize { get; set; } = 1024;
```
Not sure this should be a setting, until someone requests this. I prefer to keep the visibility of methods and properties to absolute minimum, until someone needs access.

bakgerman replied:
Removed set to 1024 in the NLogRequestPostedBodyMiddleware class

snakefoot commented on src/NLog.Web/NLogRequestPostedBodyMiddlewareConfiguration.cs:
```diff
@@ -0,0 +1,50 @@
+        /// <summary>
+        /// Defaults to UTF-8
+        /// </summary>
+        public Encoding Encoding { get; set; } = Encoding.UTF8;
```
Not sure this should be a public-setting, until someone requests this. I prefer to keep the visibility of methods and properties to absolute minimum, until someone needs access.

bakgerman replied:
Removed, set to UTF8 in the NLogRequestPostedBpdyMiddleware class

bakgerman commented:
AppVeyor seems to fail for random reasons un-related to my commits?

snakefoot commented:
Could have been nice to have HttpModule included as well, but understand how no unit-coverage and no testing-environment makes it less ideal.

Think NLogRequestPostedBodyMiddleware looks nice and simple. Would it make sense to merge it together with NLogRequestLoggingMiddleware?

I can refactor NLogRequestLoggingMiddleware, so it become optional whether to perform request-logging.

snakefoot commented:
Btw. any reason why you use `IMiddleware`, since it sounds like it has an [overhead](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/extensibility):

- Activation per client request (injection of scoped services)
- Strong typing of middleware

The `Activation` sounds like it will create an instance of the middleware for every client-request to allow constructor-injection of "scoped services")

bakgerman commented:
> Btw. any reason why you use `IMiddleware`, since it sounds like it has an [overhead](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/extensibility):
> 
> * Activation per client request (injection of scoped services)
> * Strong typing of middleware
> 
> The `Activation` sounds like it will create an instance of the middleware for every client-request to allow possible constructor-parameter-injection of "scoped services")

No, I would not like it to be constructed each time.  For now I have changed to have same structure as NLogRequestLoggingMiddleware.  I will have to fix that in the copy in my code at employer as well.  
I have changed the unit tests and I will commit that change.

bakgerman commented:
> Thank you for resolving all my comments.
> 
> Could have been nice to have HttpModule included as well, but understand how no unit-coverage and no testing-environment makes it less ideal.
> 
We can still include HttpModule version, but we must ask users for help in testing that it works and maybe should mark it experimental.  Let me know if you want it restored.

I still have additional layout renderers to PR, i went thru both Framework and Core HttpContexts to see what is not in NLog.Web, but are common in both Core and non-Core:

Request Client Certificate

Response Content Type
Response Cookies - requires Microsoft.AspNetCore.Http.Extensions NuGet in Asp.Net Core version, otherwise you can only add or delete cookie, but not read them
Response Headers

Remote Port
Local Address
Local Port

IsWebSocketRequest
WebSocketNegotiatedProtocol
WebSocketRequestedProtocols

ResponseBodyMiddlware and Layout Renderer - This one works at my employment fine, but we have <8K responses
This one, because it must write to MemoryStream, then be copied to HttpContext.Items, then memoryStream.CopyTo(actual stream back to browser or caller) always use 2x the memory.  So use carefully or we can not merge in at all.  Due to the fact that you must decide to capture BEFORE you have process the response, you cannot decide to skip response > 30K like in request body middleware where request exist BEFORE you decide to capture.

But the rest of the layout renders above are simple and just work like the rest already there.

snakefoot commented on src/Shared/LayoutRenderers/AspNetRequestPostedBodyLayoutRenderer.cs:
```diff
@@ -0,0 +1,68 @@
+            }
+#endif
+
+            if (items[NLogPostedRequestBodyKey] is string)
```
Could this be changed to:
```c#
var postedBody = items[NLogPostedRequestBodyKey] as string;
if (string.IsNullOrEmpty(postedBody))
{
   builder.Append(postedBody);
}
```
To skip a dictionary-lookup and cast-check

bakgerman replied:
Complete

snakefoot commented:
After pondering some more, then I think it is better not to merge `NLogRequestLoggingMiddleware` and `NLogRequestPostedBodyMiddleware`.

If necessary then one can make an extension method that includes both in one operation. Ex. want request-logging together with posted-body.

snakefoot commented on src/NLog.Web.AspNetCore/NLogRequestPostedBodyMiddleware.cs:
```diff
@@ -0,0 +1,133 @@
+    {
+        private readonly RequestDelegate _next;
+
+        private NLogRequestPostedBodyMiddlewareConfiguration _configuration { get; }
```
Can this become a readonly-field instead of a property ?

bakgerman replied:
Complete

bakgerman replied:
yes the other layout renders will be several small PRs after this is complete

snakefoot commented on src/NLog.Web.AspNetCore/NLogRequestPostedBodyMiddleware.cs:
```diff
@@ -0,0 +1,133 @@
+                context.Request.EnableBuffering();
+
+                // Save the POST request body in HttpContext.Items with a key of '__nlog-aspnet-request-posted-body'
+                var requestBody = await GetString(context?.Request.Body).ConfigureAwait(false);
```
Can you remove the "?" so it just becomes `context.Request.Body` ?

bakgerman replied:
Complete

snakefoot commented on src/NLog.Web.AspNetCore/NLogRequestPostedBodyMiddleware.cs:
```diff
@@ -0,0 +1,133 @@
+        private bool ShouldCaptureRequestBody(HttpContext context)
+        {
+            // Perform null checking
+            if (context.Request == null)
```
Can this become `context?.Request == null` ? (With "?")

bakgerman replied:
Complete with ?.

bakgerman replied:
maybe is better with 2 separate if statement so that InternalLogger.Debug has 2 separate messages

snakefoot commented on src/Shared/LayoutRenderers/AspNetRequestPostedBodyLayoutRenderer.cs:
```diff
@@ -0,0 +1,68 @@
+        /// <summary>
+        /// The string for the key in HttpContext.Items for the POST request body
+        /// </summary>
+        public static readonly string NLogPostedRequestBodyKey = "__nlog-aspnet-request-posted-body";
```
Could this become `internal` instead of `public` ?

snakefoot replied:
Any reason why you use a string-key with slow hashcode, instead of standard object:
```c#
internal static readonly object NLogPostedRequestBodyKey = new object();
```

bakgerman replied:
Made internal.

bakgerman replied:
Converted to object as requested.

snakefoot commented on src/NLog.Web.AspNetCore/NLogRequestPostedBodyMiddlewareConfiguration.cs:
```diff
@@ -0,0 +1,46 @@
+        /// <summary>
+        /// Defaults to true
+        /// </summary>
+        public bool DetectEncodingFromByteOrderMark { get; set; } = true;
```
Please remove this as setting, until a user needs to change it.

bakgerman replied:
Complete

snakefoot commented on src/NLog.Web.AspNetCore/NLogRequestPostedBodyMiddlewareConfiguration.cs:
```diff
@@ -0,0 +1,46 @@
+        /// The default predicate for ShouldCapture
+        /// Returns true if content length &lt;= 30KB
+        /// </summary>
+        public static bool DefaultCapture(HttpContext context)
```
Can this become a private-method (not static)? And instead of allocating a new `NLogRequestPostedBodyMiddlewareConfiguration` the active configuration?

bakgerman replied:
Complete

snakefoot commented on src/NLog.Web.AspNetCore/NLogRequestPostedBodyMiddlewareConfiguration.cs:
```diff
@@ -0,0 +1,46 @@
+        /// The maximum request size that will be captured
+        /// Defaults to 30KB
+        /// </summary>
+        public int MaximumRequestSize { get; set; } = 30 * 1024;
```
Can it be renamed to `MaxPostedBodySize` ? Maybe update comments about how it checks against `ContentLength` to protect against "very large" posted-body-payloads. And that EnableBuffering has an upper-limit of 30KByte where it changes to TEMP-files.

snakefoot commented on src/NLog.Web.AspNetCore/NLogRequestPostedBodyMiddlewareConfiguration.cs:
```diff
@@ -0,0 +1,46 @@
+        /// <summary>
+        /// The default configuration
+        /// </summary>
+        public static readonly NLogRequestPostedBodyMiddlewareConfiguration Default = new NLogRequestPostedBodyMiddlewareConfiguration();
```
Can this be changed from public to internal ?

snakefoot commented on src/NLog.Web.AspNetCore/NLogRequestPostedBodyMiddlewareConfiguration.cs:
```diff
@@ -0,0 +1,46 @@
+    /// <summary>
+    /// Contains the configuration for the NLogRequestPostedBodyMiddleware
+    /// </summary>
+    public class NLogRequestPostedBodyMiddlewareConfiguration
```
Can it be renamed to NLogRequestPostedBodyMiddlewareOptions` ?

bakgerman replied:
Complete

snakefoot commented on src/NLog.Web.AspNetCore/NLogRequestPostedBodyMiddleware.cs:
```diff
@@ -0,0 +1,136 @@
+        /// </summary>
+        /// <param name="next"></param>
+        /// <param name="configuration"></param>
+        public NLogRequestPostedBodyMiddleware(RequestDelegate next, NLogRequestPostedBodyMiddlewareOptions configuration)
```
Can input-parameter `configuration` be renamed to `options` (together with `_configuration`)

snakefoot commented:
Thank you so much for being persistent and completing all my requests.

I will try to add the ASP.NET Classic support using HttpModule, at least as first attempt

bakgerman commented:
Thank you for all of the reviews, I know you are very busy with all of the NLog work!  If you need assistance with HttpModule, please let me know.  Hopefully the code in the middleware gives a direction to develop the HttpModule.  Microsoft makes it very hard to unit tests pre-ASP.NET Core middleware (meaning HttpModule) or HttpContext class.  Maybe I can find a legacy Framework ASP.NET app at my employer that I can test with, my direct department does not but the departments we work with may have one.

snakefoot commented:
@bakgerman Would it make sense to also have a Content-Type filter, so only trying to handle posted-body when in expected-formats (Avoid images / binaries / etc.)

```
application/json
application/xml
application/xhtml
text/javascript
text/html
text/plain
text/xml
```

snakefoot commented:
@bakgerman Created #787 with a very primitive ContentType-filter.

snakefoot commented:
Would it be possible for you to update this wiki-page, and describe the recommended way for enabling `${aspnet-request-posted-body}` using middleware:

https://github.com/NLog/NLog/wiki/AspNet-Request-posted-body-layout-renderer

You are very wellcome to remove all the old noise about previous version/attempts, and just state that one must use NLog.Web.AspNetCore v5.1 (or newer).

bakgerman commented:
Certainly

snakefoot commented:
Could you also add details about using the options for the middleware?


Sent from my Galaxy



-------- Original message --------
From: Burak Akgerman ***@***.***>
Date: 18/07/2022 22:40 (GMT+01:00)
To: "NLog/NLog.Web" ***@***.***>
Cc: Rolf Kristensen ***@***.***>, State change ***@***.***>
Subject: Re: [NLog/NLog.Web] Restore aspnet-request-posted-body with middleware (PR #781)


I have changed it as requested.

—
Reply to this email directly, view it on GitHub<https://github.com/NLog/NLog.Web/pull/781#issuecomment-1188287648>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ACXZ7HDVTE27MBYDB6H3IADVUW6LHANCNFSM5WXUJUUA>.
You are receiving this because you modified the open/close state.Message ID: ***@***.***>

bakgerman commented:
Sure.

Should we add the response body middleware as example to another page?

bakgerman commented:
Added a large section about the middleware options to the end of the page

snakefoot commented:
Looks great. Thank you again.

You are very welcome to add new wiki-page about logging the response body. Then add link from request-body-layout-wiki-page.


Sent from my Galaxy



-------- Original message --------
From: Burak Akgerman ***@***.***>
Date: 18/07/2022 23:56 (GMT+01:00)
To: "NLog/NLog.Web" ***@***.***>
Cc: Rolf Kristensen ***@***.***>, State change ***@***.***>
Subject: Re: [NLog/NLog.Web] Restore aspnet-request-posted-body with middleware (PR #781)


Added a large section about the middleware options to the end of the page

—
Reply to this email directly, view it on GitHub<https://github.com/NLog/NLog.Web/pull/781#issuecomment-1188360938>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ACXZ7HHC53V2A6VDUY5PJGDVUXHIDANCNFSM5WXUJUUA>.
You are receiving this because you modified the open/close state.Message ID: ***@***.***>
